contracts/src/strealm.cairo

```diff
@@ -0,0 +1,294 @@
+                        new_address: new_reward_payer_address
+                    }
+                );
+            self.StRealm_reward_payer.write(new_reward_payer_address);
+        }
+
+
+        fn _end_latest_flow(ref self: ComponentState<TContractState>) {
+            let flow_id = self.StRealm_latest_flow_id.read();
+            let mut flow: Flow = self.StRealm_flows.read(flow_id);
+            flow.end_at = starknet::get_block_timestamp();
+            self.StRealm_flows.write(flow_id, flow);
+        }
+
+        fn _start_new_flow(ref self: ComponentState<TContractState>, new_flow_rate: u256) {
+            let new_flow_id = self.StRealm_latest_flow_id.read() + 1;
+            let new_flow: Flow = Flow { rate: new_flow_rate, end_at: BoundedInt::max() };
+            self.StRealm_flows.write(new_flow_id, new_flow);
+        }
+
+
+        fn _reset_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            let mut stream: Stream = self.StRealm_streams.read(owner);
+            stream.start_at = starknet::get_block_timestamp();
+            stream.flow_id = self.StRealm_latest_flow_id.read();
+            self.StRealm_streams.write(owner, stream);
+        }
+
+        fn _end_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            if owner.is_non_zero() {
+                let mut stream: Stream = self.StRealm_streams.read(owner);
+                stream.start_at = 0;
+                stream.flow_id = 0;
+                self.StRealm_streams.write(owner, stream);
+            }
+        }
+
+
+        fn _streamed_amount(
+            ref self: ComponentState<TContractState>,
+            token_amount: u256,
+            stream_duration: u64,
+            flow_rate: u256
+        ) -> u256 {
+            token_amount * stream_duration.into() * flow_rate
+        }
+
+
+        /// must be calld before balance updates
+        fn _claim_stream(ref self: ComponentState<TContractState>, owner: ContractAddress) {
+            if owner.is_non_zero() {
+                let stream: Stream = self.StRealm_streams.read(owner);
+                if stream.flow_id.is_non_zero() && stream.start_at.is_non_zero() {
+                    let flow: Flow = self.StRealm_flows.read(stream.flow_id);
+                    let latest_flow_id: u32 = self.StRealm_latest_flow_id.read();
+
+                    let mut stream_end_at: u64 = starknet::get_block_timestamp();
+                    if latest_flow_id > stream.flow_id {
+                        stream_end_at = flow.end_at;
+                    }
+
+                    let stream_duration = stream_end_at - stream.start_at;
+                    let erc721_component = get_dep_component!(@self, ERC721);
+
+                    let streamed_amount = self
+                        ._streamed_amount(
+                            erc721_component.balance_of(owner), stream_duration, flow.rate
```

Comment:
How do we prevent transferring NFTs around to avoid excessive claims (essentially sybil resistance)? The NFT should be staked, no? Where's that done? Does unstaking call the claim fn?

Reply:
I don't fully understand the comment so maybe you could rephrase. 
anyway, the nft isn't particularly "staked". you start accruing your reward when you delegate your realm for voting. also, the reward balance is updated any time you stop delegating or or transfer the nft.

Reply:
Aye, spotted this - in the `before_update`, every time the NFT changes ownership, the stream rewards get updated. Nicely done.